vaadin-spreadsheet-flow-parent/vaadin-spreadsheet-flow-client/src/main/java/com/vaadin/component/spreadsheet/client/js/SpreadsheetJsApi.java

```diff
@@ -463,26 +464,91 @@ public void setClass(String classNames) {
-                PopupButtonWidget widget;
-                popupButtonWidgets.put(k, widget = new PopupButtonWidget());
-                popupButtonConnectors.put(k, new PopupButtonConnector());
-                popupButtonStates.put(k, state);
-                widget.setCol(state.col);
-                widget.setRow(state.row);
-                widget.setPopupHeaderHidden(state.headerHidden);
-                widget.setSheetWidget(spreadsheetWidget.getSheetWidget(),
-                        DivElement.as(spreadsheetWidget.getSheetWidget()
-                                .getElement()));
-                widget.setPopupWidth(state.popupWidth);
-                widget.setPopupHeight(state.popupHeight);
-                spreadsheetWidget.addPopupButton(widget);
+    private static String getPopupKey(String sheet, int row, int column) {
+        return sheet + "_" + row + "_" + column;
+    }
+
+    public void addPopupButton(String rawState) {
+        PopupButtonState state = Parser.parsePopupButtonState(rawState);
+        String key = getPopupKey(state.sheet, state.row, state.col);
+
+        PopupButtonWidget widget;
+
+        if (!popupButtonWidgets.containsKey(key)) {
+            PopupButtonConnector connector;
+            popupButtonConnectors.put(key,
+                    connector = new PopupButtonConnector());
+            popupButtonWidgets.put(key, widget = connector.getWidget());
+
+            popupButtonStates.put(key, state);
+            connector.setPopupButtonServerRpc(getServerRpcInstance());
+            connector.init();
+            widget.setCol(state.col);
+            widget.setRow(state.row);
+            widget.setPopupHeaderHidden(state.headerHidden);
+            widget.setSheetWidget(spreadsheetWidget.getSheetWidget(), DivElement
+                    .as(spreadsheetWidget.getSheetWidget().getElement()));
+            widget.setPopupWidth(state.popupWidth);
+            widget.setPopupHeight(state.popupHeight);
+        } else {
+            widget = popupButtonWidgets.get(key);
+        }
+        spreadsheetWidget.addPopupButton(widget);
+    }
+
+    public void removePopupButton(String rawState) {
+        PopupButtonState state = Parser.parsePopupButtonState(rawState);
+        String key = getPopupKey(state.sheet, state.row, state.col);
+        PopupButtonWidget popupButton = popupButtonWidgets.get(key);
+
+        if (popupButton != null) {
+            spreadsheetWidget.removePopupButton(popupButton);
+            popupButtonWidgets.remove(key);
+            popupButtonConnectors.remove(key);
+            popupButtonStates.remove(key);
+        }
+    }
+
+    public static class ContentWidget extends Widget {
+        public ContentWidget(Element contentElement) {
+            setElement(contentElement);
+        }
+    }
+
+    private static native Element getPopupContentContainer(
+            String contentParentId) /*-{
+        if (!(Vaadin && Vaadin.Flow && Vaadin.Flow.clients.ROOT)) {
```

Comment:
This fails when using a XSI GWT linker like in SDM.

Try this code with the SDM patch (cherry-picking the last commit in https://github.com/vaadin/flow-components/commits/sdm branch), then run the demo in SDM and you will get the exception:

```
FlowClient.js?2d7e:196 Uncaught ReferenceError: Vaadin is not defined
    at getPopupContentContainer_0_g$ (SpreadsheetJsApi.java:519:1)
    at SpreadsheetJsApi_1_g$.onPopupButtonOpened_0_g$ [as onPopupButtonOpened] (SpreadsheetJsApi.java:528:1)
    at VaadinSpreadsheet.onPopupButtonOpen (vaadin-spreadsheet.js?ffdc:407:1)
    at Object.eval (eval at Zt (FlowClient.js?2d7e:1:1), <anonymous>:3:11)
    at Zt (FlowClient.js?2d7e:1064:164)
```

Reply:
I added a commit to the branch fixing this.

Reply:
Thanks for taking care of this issue 🙇🏻‍♂️